src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -3977,24 +4185,38 @@ ColocationGroupCreateCommandList(void)
 	 * However, it is useful to get consistent behavior, both for regression
 	 * tests and also in production systems.
 	 */
-	SysScanDesc scanDescriptor =
-		systable_beginscan_ordered(pgDistColocation, colocationIdIndexRel,
-								   NULL, 0, NULL);
-
-	HeapTuple colocationTuple = systable_getnext_ordered(scanDescriptor,
-														 ForwardScanDirection);
+	SysScanDesc scanDescriptor = systable_beginscan_ordered(pgDistColocation,
+															colocationIdIndexRel,
+															NULL, 0, NULL);
 
-	while (HeapTupleIsValid(colocationTuple))
+	MemoryContext oldContext = MemoryContextSwitchTo(context->context);
+	HeapTuple nextTuple = NULL;
+	while (true)
 	{
-		if (hasColocations)
+		if (!InMetadataSyncTestMode(context))
+		{
+			MemoryContextReset(context->context);
+		}
+
+		/* index scan in oldContext. Or, we get SEGFAULT as we reset metadataSyncContext */
+		MemoryContextSwitchTo(oldContext);
```

Comment:
As discussed, we don't need `systable_beginscan_ordered` anymore. We needed in the past because `WITH` queries that inserted multiple objects was not ordered in the tests outputs. Now, each object is a single `WITH` command -- which we should consider converting to a non-WITH clause.

If we avoid `systable_beginscan_ordered`, we can avoid these two unusual `MemoryContextSwitchTo` calls

Reply:
After internal discussion, we decided to leave `WITH` commands as we may decide to batch commands in future for performance improvements.